platform/api/oci/signer.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	genericHeaders = []string{
+		"date",
+		"(request-target)",
+		"host",
+	}
+
+	bodyHeaders = []string{
+		"content-length",
+		"content-type",
+		"x-content-sha256",
+	}
+
+	requiredHeaders = map[string][]string{
+		"get":    genericHeaders,
+		"head":   genericHeaders,
+		"delete": genericHeaders,
+		"put":    append(genericHeaders, bodyHeaders...),
+		"post":   append(genericHeaders, bodyHeaders...),
+	}
+)
+
+type Signer struct {
+	key *rsa.PrivateKey
+
+	// <tenancy>/<user>/<rsa fingerprint>
+	keyID string
+}
+
+func New(privateKey *rsa.PrivateKey, keyID string) *Signer {
+	return &Signer{
+		key:   privateKey,
+		keyID: keyID,
+	}
+}
+
+func (s *Signer) SignRequest(req *http.Request) error {
+	var reqHeaders []string
+	if val, ok := requiredHeaders[strings.ToLower(req.Method)]; ok {
+		reqHeaders = val
+	} else {
+		return fmt.Errorf("unknown method")
+	}
+
+	verb := strings.ToLower(req.Method)
+	signBody := verb == "post" || verb == "put"
+	err := s.injectMissingHeaders(req, signBody)
+	if err != nil {
+		return fmt.Errorf("injecting headers: %v", err)
+	}
+
+	signable, err := s.generateMessage(reqHeaders, req)
+	if err != nil {
+		return fmt.Errorf("generating signature message: %v", err)
+	}
+
+	signature, err := s.sign(signable)
+	if err != nil {
+		return fmt.Errorf("signing message: %v", err)
+	}
+
+	signatureTemplate := s.buildSignatureTemplate(reqHeaders)
+	req.Header.Set("Authorization", fmt.Sprintf(signatureTemplate, signature))
```

Comment:
Building a format string by concatenating user-supplied data is generally considered a bad idea.  (It violates memory safety in C, and in not-C it still risks unintended interpretation of user-controlled data.)  Why not just pass the signature into `buildSignatureTemplate`?